shared/js/places.js

```diff
@@ -0,0 +1,103 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+  'places' DataStore wrapper
+  Ensures the data format of the 'places' DataStore stays consistent.
+**/
+
+'use strict';
+
+/* exported getPlacesStore */
+
+var getPlacesStore = (() => {
+  var _store;
+  var verify = (obj) => {
+    var field;
+    var obligatory = {
+      url: 'string'
+    };
+    var allowed = {
+      url: 'string',
+      title: 'string',
+      icons: 'object',
+      frecency: 'number',
+      visits: 'object', // Array, to be specific
+      screenshot: 'object',
+      fxsyncId: 'string',
+      visited: 'number'
+    };
+    if (typeof obj !== 'object') {
+      throw new Error(`Records in places DataStore should be objects.`);
+    }
+    for (field in obligatory) {
+      if (typeof obj[field] !== obligatory[field]) {
+        throw new Error(`Missing obligatory field ${field} in places DataStore \
+record.`);
+      }
+    }
+    for (field in obj) {
+      if (typeof obj[field] !== allowed[field]) {
+        throw new Error(`Field ${field} of type ${typeof obj[field]} not allowe\
+d in places DataStore record.`);
+      }
+    }
+    // Extra specific check for visits as descending timestamps array:
+    if (obj.visits) {
+      if (!Array.isArray(obj.visits)) {
+        throw new Error(`Field visits should be an Array or undefined in places\
+ DataStore record.`);
+      }
+      let last = 99999999999; // 16 November in the year 5138
+      for (let i = 0; i < obj.visits.length; i++) {
+        if (typeof obj.visits[i] !== 'number') {
+          throw new Error(`Timestamps in visits Array should be numbers in plac\
+es DataStore record.`);
+        }
+        if (obj.visits[i] > last) {
+          throw new Error(`Timestamps in visits Array should be in seconds and \
+descending in places DataStore record.`);
+        }
+        last = obj.visits[i];
+      }
+    }
+  };
+
+  return () => {
+    if (_store) {
```

Comment:
This type of caching is racy

getPlacesStore();
getPlacesStore();

is going to open 2, I would either not both caching or use a cache that wont race ie

```
  if (_store) { 
    return _store;
  }
  _store = navigator.getDataStores('place').then(stores => 
    return { .... };
  });
  return _store;